website/blog/deployers.md

```diff
@@ -0,0 +1,610 @@
+</div>
+
+### Components
+
+First, we implement `ActiveComponent`. When a component hosted by a weavelet
+calls `weaver.Get[T]` for some component `T`, the `ActivateComponent` method
+gets invoked to activate `T`. `ActivateComponent` should start the
+component&mdash;potentially with multiple replicas&mdash;if it hasn't already
+been started.
+
+Our handler calls `deployer.spawn` to spawn the weavelet hosting the requested
+component, it it hasn't already been spawned. The handler then calls
+`UpdateRoutingInfo` to inform the weavelet of the requested component's address.
+This allows the weavelet to perform RPCs with the component.
+
+A deployer should call `UpdateRoutingInfo` whenever there is a change to the
+routing information of a component for which a weavelet has called
+`ActivateComponent`. For example, if a deployer detects that a weavelet hosting
+component `A` has crashed, it should call `UpdateRoutingInfo` on all weavelets
+that have called `ActivateComponent` on `A` with new routing information that
+omits the address of the failed weavelet.
+
+```go
+// Responsibility 1: Components.
+func (h *handler) ActivateComponent(_ context.Context, req *protos.ActivateComponentRequest) (*protos.ActivateComponentReply, error) {
+    // Spawn the component, if it hasn't already been spawned.
+    w, err := h.deployer.spawn(req.Component)
+    if err != nil {
+        return nil, err
+    }
+
+    // Tell the weavelet the address of the requested component.
+    h.envelope.UpdateRoutingInfo(&protos.RoutingInfo{
+        Component: req.Component,
+        Replicas:  []string{w.address},
+    })
+    return &protos.ActivateComponentReply{}, nil
+}
+```
+
+<div class="note">
+<code>weaver multi</code>, like our deployer, spawns weavelets in subprocesses. <code>weaver gke</code> spawns weavelets in <a href="https://kubernetes.io/docs/concepts/workloads/controllers/deployment/">Kubernetes deployments</a>.
+</div>
+
+### Listeners
+
+Next, we implement the listener methods. When a component requests a network
+listener, the `GetListenerAddress` method is invoked. This method returns the
+address on which the component should listen. Our simple deployer always returns
+`"localhost:0"`. After the requesting weavelet receives this address, it creates
+a network listener on the address and invokes the `ExportListener` method with
+the concrete address that it's listening on. For example, `"localhost:0"` could
+be resolved to `"127.0.0.1:35879"`.
+
+```go
+// Responsibility 2: Listeners.
+func (h *handler) GetListenerAddress(_ context.Context, req *protos.GetListenerAddressRequest) (*protos.GetListenerAddressReply, error) {
+    return &protos.GetListenerAddressReply{Address: "localhost:0"}, nil
+}
+
+func (h *handler) ExportListener(context.Context, *protos.ExportListenerRequest) (*protos.ExportListenerReply, error) {
```

Comment:
It's not clear to me what we are trying to convey with the method definition. Given the preceding text, I was hoping to see a ExportListener call with "127.0.0.1:35879".

Reply:
This was unclear. I updated the code and expanded the explanation. `ExportListener` will get called with `"127.0.0.1:35879"` by the weavelet, but the deployer doesn't call `ExportListener` itself.